purchase_sale_inter_company/models/purchase_order.py

```diff
@@ -0,0 +1,190 @@
+        :param name : the origin client reference
+        :rtype name : string
+        :param partner : the partner reprenseting the company
+        :rtype partner : res.partner record
+        :param dest_company : the company of the created SO
+        :rtype dest_company : res.company record
+        :param direct_delivery_address : the address of the SO
+        :rtype direct_delivery_address : res.partner record
+        """
+        self.ensure_one()
+        delivery_address = direct_delivery_address or partner or False
+        new_order = self.env["sale.order"].new(
+            {
+                "company_id": dest_company.id,
+                "client_order_ref": name,
+                "partner_id": partner.id,
+                "date_order": self.date_approve,
+                "auto_purchase_order_id": self.id,
+            }
+        )
+        for onchange_method in new_order._onchange_methods["partner_id"]:
+            onchange_method(new_order)
+        new_order.user_id = False
+        if delivery_address:
+            new_order.partner_shipping_id = delivery_address
+        if self.notes:
+            new_order.note = self.notes
+        new_order.commitment_date = self.date_planned
+        return new_order._convert_to_write(new_order._cache)
+
+    def _prepare_sale_order_line_data(self, purchase_line, dest_company, sale_order):
+        """Generate the Sale Order Line values from the PO line
+        :param purchase_line : the origin Purchase Order Line
+        :rtype purchase_line : purchase.order.line record
+        :param dest_company : the company of the created SO
+        :rtype dest_company : res.company record
+        :param sale_order : the Sale Order
+        """
+        new_line = self.env["sale.order.line"].new(
+            {
+                "order_id": sale_order.id,
+                "product_id": purchase_line.product_id.id,
+                "product_uom": purchase_line.product_uom.id,
+                "product_uom_qty": purchase_line.product_qty,
+                "auto_purchase_line_id": purchase_line.id,
+                "display_type": purchase_line.display_type,
+            }
+        )
+        for onchange_method in new_line._onchange_methods["product_id"]:
+            onchange_method(new_line)
+        new_line.update({"product_uom": purchase_line.product_uom.id})
+        if new_line.display_type in ["line_section", "line_note"]:
+            new_line.update({"name": purchase_line.name})
+        return new_line._convert_to_write(new_line._cache)
+
+    def button_cancel(self):
+        sale_orders = (
+            self.env["sale.order"]
+            .sudo()
+            .search([("auto_purchase_order_id", "in", self.ids)])
+        )
+        for so in sale_orders:
+            if so.state not in ["draft", "sent", "cancel"]:
+                raise UserError(
+                    self.env._("You can't cancel an order that is %s") % so.state
+                )
```

Comment:
In the translation function, you can pass the arguments as parameters.

```suggestion
                raise UserError(
                    self.env._("You can't cancel an order that is %s", so.state)
                )
```